Migrating json metadata plugin to use filtering engine #3750

robbarnsley opened the pull request:
<!-- Please read https://github.com/rucio/rucio/blob/master/CONTRIBUTING.rst before submitting a pull request -->

joeldierkes commented on bin/rucio:
```diff
@@ -676,26 +664,47 @@ def list_dids_extended(args):
+                return FAILURE
+
+    try:
+        filters, type_ = parse_did_filter_from_string_fe(args.filter, name)
```
What if `--filter` is not provided? It's optional

robbarnsley replied:
`args.filter` defaults to None if not specified as an argument, it's then handled by the function

joeldierkes commented on lib/rucio/core/did_meta_plugins/__init__.py:
```diff
@@ -16,186 +16,211 @@
+        for metadata_plugin in METADATA_PLUGIN_MODULES:
+            if metadata_plugin.get_plugin_name().lower() == plugin.lower():
+                return metadata_plugin.get_metadata(scope, name, session=session)  
+    raise NotImplementedError('Metadata plugin "%s" is not enabled on the server.' % plugin)    #FIXME: this exception isn't returned correctly?
```
I don't understand the `FIXME`, is it not correctly handled?

Also, please document that an exception is thrown (`:raises ...: ...`). It was forgotten before.

robbarnsley replied:
fixme removed. added doc re: raises

joeldierkes commented on lib/rucio/core/did.py:
```diff
@@ -1606,7 +1606,7 @@ def list_dids(scope, filters, did_type='collection', ignore_case=False, limit=No
 @read_session
 def list_dids_extended(scope, filters, did_type='collection', ignore_case=False, limit=None,
-                       offset=None, long=False, recursive=False, session=None):
+                       offset=None, long=False, recursive=False, ignore_dids=None, session=None):
```
A test would be nice, have a look at https://github.com/rucio/rucio/blob/c857f58149b09d785338c0abe02e3487b85f4cd9/lib/rucio/tests/test_did_meta_plugins.py

robbarnsley replied:
added, but not here (to filter engine)

joeldierkes commented on lib/rucio/client/didclient.py:
```diff
@@ -94,28 +97,38 @@ def list_dids(self, scope, filters, did_type='collection', long=False, recursive
-        payload['type'] = did_type
-        payload['recursive'] = recursive
+        # stringify dates.
+        if isinstance(filters, dict):   # backwards compatability for filters as single {}
```
A test would be good if `filters` is an array of dictionaries.

joeldierkes reviewed (changes requested):
The PR is way to big. It is a best practice to make commits rather small and descriptive with one objective. They are easier to review and less likely to introduce a bug. Feel free to open multiple PRs time after time and introduce features one after another.

Also, your commit messages are not descriptive at all. "intermediate commit" is not informative.

robbarnsley commented:
My mistake - this should've been a draft (hence the FIXMEs, unsquashed commits and zero tests). Marked it as such now.

Just to note, I know what best practice is - it is a little difficult for this feature because it spans multiple layers of the architecture. It would be difficult to compartmentalise.

bari12 commented on lib/rucio/core/did_meta_plugins/__init__.py:
```diff
@@ -16,186 +16,211 @@
+# By default, the custom metadata plugin is set to the prepackaged json module. This utilises a separate table, 
+# "did_meta", allowing users to add custom metadata key/value pairs of any type.
+#
+HARDCODED_METADATA_PLUGIN_MODULE_PATH = "rucio.core.did_meta_plugins.did_column_meta.DidColumnMeta"
```
Maybe rename this, I know it doesn't come from you, but I was never a fan of the HARDCODED and FALLBACK naming.

robbarnsley replied:
How about 'base' and 'extra', or 'base' and 'custom'?

Side question: I presume it is critical to core functionality to have the metadata functions always look for metadata in the did columns first? Reason I ask is that, as it stands, there is no way to have a key with the same name as one in the did column store in a "custom" metadata store. 

It's also, in my opinion, not very transparent. If I was defining a list of metadata plugins in the `metadata -> plugins` section of the rucio cfg, i'd expect it to only use those plugins, and in the order I put them. That field would be better called something like `extra_plugins`.

bari12 replied:
We mostly did it this way to preserve the functionality of how Rucio was working before (--> column information is always queried first)
But I agree, this is not very intuitive/transparent. Best would probably be to just go to an ordered list, and metadata plugins are queried in that order. The rucio default should be to do column first, but it doesn't have to be like that for every community.

robbarnsley committed:
Ported filter engine to work with JSON metadata.

Functionality is now provided for all sqla databases that support JSON excluding Oracle.

Amendments include:

- Refactored metadata __init__.py (added comments and changed variable names).
- Refactored metadata plugin did_column_meta.py (HARDCODED_KEYS now pulled from models.DataIdentifier)
- Switched list_dids_extended to use filter engine for did column plugin.
- Amended org schemas as filters is now a nested dict and not just a dict.
- Added unit tests.

robbarnsley commented on lib/rucio/core/did_meta_plugins/__init__.py:
```diff
@@ -16,157 +16,194 @@
 # - Aristeidis Fkiaras <aristeidis.fkiaras@cern.ch>, 2020
 # - Benedikt Ziemons <benedikt.ziemons@cern.ch>, 2020-2021
+# - David Población Criado <david.poblacion.criado@cern.ch>, 2021
+# - Rob Barnsley <robbarnsley@users.noreply.github.com>, 2021-2022
```
A lot of the changes in this file are just housekeeping - changing variable names, adding comments and amending docstrings etc.

robbarnsley commented on lib/rucio/core/did_meta_plugins/__init__.py:
```diff
@@ -16,157 +16,194 @@
+
+# Set restricted character set for metadata in form character: reason
+#
+RESTRICTED_CHARACTERS = {
```
Disallow certain characters in metadata. '.' is used as a delimiter between the key and operator.

robbarnsley commented on lib/rucio/core/did_meta_plugins/did_column_meta.py:
```diff
@@ -14,105 +14,44 @@
 from rucio.db.sqla.constants import DIDType
 from rucio.db.sqla.session import stream_session, read_session, transactional_session
 
-HARDCODED_KEYS = [
```
Removed this in favour of getting the list directly from the model definition, with some exceptions.

robbarnsley commented on lib/rucio/core/did_meta_plugins/did_column_meta.py:
```diff
@@ -241,128 +180,141 @@ def list_dids(self, scope, filters, did_type='collection', ignore_case=False, li
+                (models.DataIdentifier.scope, operator.eq, scope),
+                (models.DataIdentifier.suppressed, operator.ne, true())
+            ]
+        )
```
just copying over the existing core `list_dids` function (which uses the filter engine).

robbarnsley commented on lib/rucio/core/did_meta_plugins/filter_engine.py:
```diff
@@ -235,72 +214,121 @@ def _translate_filters(self, model_class, strict_coerce=True):
         self._filters = filters_translated
         return list(mandatory_model_attributes)
 
+    def _try_typecast_string(self, value):
+        """
+        Check if string can be typecasted to bool, datetime or float.
+
+        :param value: The value to be typecasted.
+        :returns: The typecasted value.
+        """
+        value = value.replace('true', 'True').replace('TRUE', 'True')
+        value = value.replace('false', 'False').replace('FALSE', 'False')
+        for format in VALID_DATE_FORMATS:       # try parsing multiple date formats.
+            try:
+                value = datetime.strptime(value, format)
+            except ValueError:
+                continue
+            else:
+                return value
+        try:
+            value = ast.literal_eval(value)     # will catch float, int and bool
+        except (ValueError, SyntaxError):
+            pass
+        return value
+
```
When user inputs a metadata filter via CLI, need some way of attempting to cast the value of each key from string (important later when constructing the sqla filter expression).

robbarnsley commented on lib/rucio/core/did_meta_plugins/filter_engine.py:
```diff
@@ -235,72 +214,121 @@ def _translate_filters(self, model_class, strict_coerce=True):
+                    if oper == operator.ne:                                                         # set .ne operator to include NULLs.
+                        expression = or_(expression, key.is_(None))
+                elif json_column:                                                                   # -> this key filters on the content of a json column
+                    if session.bind.dialect.name == 'oracle':
+                        # Compiling for Oracle yields a UnsupportedCompilationError so fall back to support for equality only.
+                        # TODO
+                        if oper != operator.eq:
+                            raise exception.FilterEngineGenericError("Oracle implementation does not support this operator.")
+                        if isinstance(value, str) and any([char in value for char in ['*', '%']]):
+                            raise exception.FilterEngineGenericError("Oracle implementation does not support wildcards.")
+                        expression = text("json_exists(meta,'$?(@.{} == \"{}\")')".format(key, value))
```
yet to implement this for Oracle - requires a more bespoke filter expression construction

robbarnsley commented on lib/rucio/core/did_meta_plugins/filter_engine.py:
```diff
@@ -317,3 +345,66 @@ def evaluate(self):
+                    filters += ' AND '
+            if or_group != self._filters[-1]:
+                filters += ' OR\n'
+        return filters
```
not new function, just moved

robbarnsley commented on lib/rucio/tests/test_filter_engine.py:
```diff
@@ -166,8 +186,43 @@ def test_OperatorsEqualNotEqual(self, session=None):
 
+        # Plugin: JSON
+        #
+        if json_implemented(session=session):
+            did_name1 = self._create_tmp_DID()
+            did_name2 = self._create_tmp_DID()
+            did_name3 = self._create_tmp_DID()
+            set_metadata(scope=self.tmp_scope, name=did_name1, key='testkeyint1', value=1)
+            set_metadata(scope=self.tmp_scope, name=did_name2, key='testkeyint2', value=2)
+            set_metadata(scope=self.tmp_scope, name=did_name3, key='testkeyint3', value=2)
+
+            dids = []
+            q = FilterEngine('testkeyint1=1', model_class=models.DidMeta, strict_coerce=False).create_sqla_query(
+                additional_model_attributes=[
+                    models.DidMeta.scope,
+                    models.DidMeta.name
+                ],
+                json_column=models.DidMeta.meta)
+            dids += [did for did in q.yield_per(5)]
+            dids = set(dids)
+            self.assertEqual(list(map(lambda did: did.name in (did_name1, did_name2, did_name3), dids)).count(True), 1)
```
equality supported by all sqla flavours, rest of the cases are not supported by oracle yet

robbarnsley commented:
I recognise this is a rather sizable PR, and there are a few housekeeping bits that could've probably been packaged as a separate request, however a lot of the changes here are all required for the JSON plugin to work with the filter engine.

From this point, the core functionality should now exist and future PRs will be smaller and more focused.

I've added a few comments that should hopefully guide a review.

bari12 commented on lib/rucio/core/did_meta_plugins/json_meta.py:
```diff
@@ -150,41 +139,77 @@ def delete_metadata(self, scope, name, key, session=None):
+                        yield {
+                            'scope': did.scope,
+                            'name': did.name,
+                            'did_type': 'Not available with {} plugin.'.format(self.get_plugin_name()),
```
I think it would be better to set these to `None`. Just to make sure this isn't accidentally processed otherwise.

robbarnsley replied:
Yep, no problem.

bari12 commented on lib/rucio/core/did_meta_plugins/json_meta.py:
```diff
@@ -150,41 +139,77 @@ def delete_metadata(self, scope, name, key, session=None):
+        if limit:
+            query = query.limit(limit)
+        if recursive:
+            from rucio.core.did import list_content
```
This can't be loaded at top? Circular dependency?

robbarnsley replied:
Not sure of the reasoning (someone else did this). Doesn't look circular but get a bunch of import errors when it's at the top.

bari12 reviewed (commented):
Hi @robbarnsley 
This is very nice work. Thanks a lot. 
I have two minor comments, otherwise I think this is GTG.

Maybe one recommendation for the future in terms of large PRs. It would help to separate things into different contained commits. Similar to the comments you put in the PR later.
E.g. one commit about just moving code around; One commit about housekeeping; one commit about adding a new feature
I know it's more work to do that, but makes the review much easier for large PRs like this.

robbarnsley committed:
Merge branch 'master' into feature-3750-Migrating_json_metadata_plugin_to_use_filtering_engine

robbarnsley commented:
> Maybe one recommendation for the future in terms of large PRs. It would help to separate things into different contained commits. Similar to the comments you put in the PR later.

Fair point, will keep this in mind.
